M2-5579: [Participant Overview] List + Sort

sultanofcardio opened the pull request:
### 📝 Description

🔗 [<img width="12" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/314e04f9-eb5a-492d-9a3b-3294f5b53360"> Jira Ticket M2-5579](https://mindlogger.atlassian.net/browse/M2-5579)
🔗 [<img width="8" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/ae08667b-7755-40bc-b08d-db0a729812d0"> Figma link](https://www.figma.com/file/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?type=design&node-id=808%3A126146&mode=design&t=NoVlk8x5mTM773pW-1)

This PR introduces the `/dashboard/<applet-id>/participants` route with the new design for the participants overview page. Continuing the theme of copy over modify, I decided to introduce a new `<Participants>` feature component for this change.

### 📸 Screenshots

#### Before

![image](https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/0e84d773-5e7b-4004-91c0-220dc2ece532)

#### After

![image](https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/d57bc2f2-cd32-40df-ba01-9902f9c6f35a)

### 🪤 Peer Testing

> [!NOTE]
> You'll need at least one applet for this test

1. Add this variable to your `.env` file
  ```shell
  REACT_APP_FEATURE_FLAG_APPLET_MULTI_INFORMANT=true
  ```
2. Start the backend API (from the [feature-multiinformant](https://github.com/ChildMindInstitute/mindlogger-backend-refactor/tree/feature-multiinformant) branch!)
3. Start the Admin API and navigate to the Applet Participants overview tab

https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/af054c8c-f521-4de4-99b5-ed704d6d32cd

4. Add participants if necessary

Here are some criteria to keep in mind when testing this change:
- If I access the participants tab, a list of participants loads for viewing
- This list should include essential details such as: ID, NickName, Latest Activity, and Schedule
- If study information is missing, it's marked with a "–".
- Clicking the …. on the list exposes the ability to:
  - edit participant, updated to full account, when applicable 
  - export data
  - remove participant
  - assign activity
- When I am viewing a table with multiple columns and I click on the header of a column the table should sort the rows based on the values in that column in ascending order.
  - If I click the column header again, the order should toggle to descending.
- When a column in a table is sorted, a visual indicator arrow should appear next to the column header.
- The direction of the arrow indicates the sort order: an up arrow (↑) for ascending order and a down arrow (↓) for descending order.

### ✏️ Notes

#### Table Headers

I opted to leave the bottom border of the table header visible because the table has two header rows (as opposed to the design which only has one), and I wasn't able to find a combination of header borders that didn't feel weird besides what was already there.

sultanofcardio committed:
M2-5579: Participant Overview Table WIP

sultanofcardio committed:
M2-5579: Fix linting issues

sultanofcardio committed:
M2-5579: Undelete husky files

sultanofcardio committed:
M2-5579: Remove custom table and clean up styles

sultanofcardio committed:
M2-5579: Add missing French translations 😅

farmerpaul commented on src/resources/app-fr.json:
```diff
@@ -674,6 +676,7 @@
   "lastName": "Nom de Famille",
   "lastNameRequired": "Le nom de famille est requis",
   "latestActive": "Dernière activité",
+  "latestActivity": "Dernère Activité",
```
```suggestion
  "latestActivity": "Dernière Activité",
```

sultanofcardio replied:
I'm a bit confounded with this one. When I translate "Latest Activity" on Google Translate, I get Dernère (which was apparently marked as correct by contributors):

<img width="1315" alt="Google Translate translating Latest Activity in English to Dernère Activité in French" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/82251f06-0d24-4b88-a6c9-94d4295ae27d">

However, when I try to translate it back it tries to correct me as you did from Dernère to Dernière:

<img width="1315" alt="Google Translate translating Dernère Activité in French to Last Activity in English, suggesting that I actually meant Dernière Activité (which I didn't)" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/43bc455b-409c-4bf2-8926-8cab8ad245be">

Given that there's an existing entry for `"latestActive": "Dernière activité"`, I wonder which one of these is correct 🤔

farmerpaul replied:
I guess I don't trust Google translate's reviewers here. Googling "dernère vs dernière" doesn't return any results suggesting "dernère" is a valid spelling of the word, though some people have indeed misspelled it online. In my 8+ years of French immersion I never encountered a version of that word that omits the "i".

As for the other key, it's possible that the translation there isn't quite right (maybe something like "Dernier actif" could work), or that they should both be the same even though in English there's meaningful distinction (French can be weird that way).

Having said all that, IMHO it shouldn't be the responsibility of non-bilingual engineers to be making these decisions… given that Google Translate has been the go-to till now, probably many Google-translated keys aren't correct and need to pass through a proper translation service.

sultanofcardio replied:
That's fair. I will make the change as I trust you're much better with French that I am 😅

farmerpaul commented on src/modules/Dashboard/features/Participants/Participants.tsx:
```diff
@@ -0,0 +1,564 @@
+        width: ParticipantsColumnsWidth.Pin,
+      },
+      pin: {
+        content: () => <Pin isPinned={isPinned} data-testid="dashboard-respondents-pin" />,
```
Just making a note of this here (though you don't need to change this line):

The `<Pin>`'s default (unpinned) colour is out of sync with Figma:
- It should be `outline_variant` when unpinned (instead of `surface_variant`) – see `Pin.styles.tsx`
- The value for the `outline_variant` token is also not in sync with what's in Figma; it should be `#c2c7cf` (instead of `#c4c9d0`) – see `palette.ts`

Might seem minor but I couldn't help but clock its lack of contrast with the background when you're hovering over a table row (my a11y-attuned eyes I guess 👀).

sultanofcardio replied:
Great catch Paul! I hesitate to make this change as part of this ticket since the `Pin` component is used in a few places, and `outline_variant` is used in even more places. I'd rather make that change as part of a ticket where I can investigate the usages

What do you think about overriding the unpinned style for this specific usage?

farmerpaul replied:
Hmm… I think it's probably best to leave this to a separate ticket rather than fragment this use-case, which I think will introduce unneeded complexity.

sultanofcardio replied:
Ok no prob, I'll create the Jira ticket. I ran into a hiccup and asked about it in Slack but I'll follow up with its creation once we iron that out

sabrichu commented on src/svgSprite.ts:
```diff
@@ -293,6 +293,12 @@ export const svgSprite = (): string => `
 
+    <symbol viewBox="0 0 24 24" id="svg-dropdown-down-outlined">
+      <path d="M16.9997 9.1697C16.8123 8.98345 16.5589 8.87891 16.2947 8.87891C16.0305 8.87891 15.7771 8.98345 15.5897 9.1697L11.9997 12.7097L8.4597 9.1697C8.27234 8.98345 8.01889 8.87891 7.7547 8.87891C7.49052 8.87891 7.23707 8.98345 7.0497 9.1697C6.95598 9.26266 6.88158 9.37326 6.83081 9.49512C6.78004 9.61698 6.75391 9.74769 6.75391 9.8797C6.75391 10.0117 6.78004 10.1424 6.83081 10.2643C6.88158 ...
+      fill=${variables.palette.on_surface_variant || 'currentColor'} />
```
I realize this is a pattern in the rest of this file, but I feel like we should always be using `currentColor` in these sprites and leave it up to the consumers of the `<Svg />` component to decide the fill, rather than having the sprite have a default color that would need to be overwritten

farmerpaul replied:
+1

sultanofcardio replied:
Sure, I think the result would be the same so I'm happy to break the pattern here

farmerpaul commented on src/modules/Dashboard/features/Participants/Participants.tsx:
```diff
@@ -0,0 +1,564 @@
+    () => setIsLoading(false),
+  );
+
+  const handlePinClick = ({ respondentId, subjectId }: HandlePinClick) => {
```
Placing this comment here, as it's related to the pin interaction.

I get that optimistic updates are probably not in scope for now (even though that would be ideal 😏). For now, just wanted to clock that the loading state doesn't seem to be fully aligned with the toggling of the pin icon – the loader disappears, then there's a short delay, and then the UI updates:

https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/8e55428a-333c-4553-80e8-01363d1e39e8

I don't know if it's easy to fix this, but just wanted to flag.

sultanofcardio replied:
Looks like this is an easy one. I'll go for the quick win for now, as doing an optimistic pin update isn't as straightforward as it seems

farmerpaul reviewed (commented):
Looks great @sultanofcardio! I just noticed a few pretty minor things, and I'm curious what your thoughts are on the pinning action loading state. Nice work!

sultanofcardio committed:
M2-5579: Remove defaulting symbol fill to `variables.palette.on_surface_variant`

This is an existing pattern that I followed when adding this icon, but having the consumer explicitly specify the color is likely a better approach

sultanofcardio committed:
M2-5579: Fix pinning delay after spinner disappears

It appears that the function that handles pinning was prematurely turning off the loading spinner before the data in the table could be reloaded, which is what effects the pin action.

The pin action is a combination of the pin icon being filled, and the row being moved to the top of the table and sorted among the pinned items

IhorNesterenko99 commented on src/modules/Dashboard/features/Participants/Participants.styles.ts:
```diff
@@ -0,0 +1,86 @@
+  }
+`;
+
+export const MuiCheckbox = styled(Checkbox)``;
```
It seems this row can be removed.

sultanofcardio replied:
Yup, thanks for pointing this out 👍

IhorNesterenko99 commented on src/modules/Dashboard/features/Participants/Participants.utils.tsx:
Could you please add test coverage?

sultanofcardio replied:
Done

IhorNesterenko99 commented on src/modules/Dashboard/features/Participants/Participants.tsx:
Could you please add test coverage?

sultanofcardio replied:
Done

sultanofcardio committed:
M2-5579: Remove unnecessary styled MuiCheckbox component

sultanofcardio committed:
M2-5579: Update empty state copy

sultanofcardio committed:
M2-5579: Add some tests

sultanofcardio committed:
M2-5579: Update French translation of "Latest Activity"

farmerpaul commented on src/modules/Dashboard/features/Participants/Participants.styles.ts:
```diff
@@ -0,0 +1,84 @@
+  td,
+  th {
+    min-width: 130px;
+  }
```
Just noting that many lengths in this file are specified in `px`, but if you search the code, `rem` is primarily used throughout the app. Whether or not it will currently impact anything is uncertain, but the beauty of `rem` is that you can use it to scale the whole site via a root-level `font-size` change (if that's ever desired for a given breakpoint, perhaps in the future – handy for widescreen breakpoints, for instance). The conversion to `rem` is easy since `10px == 1rem`.

sultanofcardio replied:
Good call. I think that's a worthwhile change. Thanks for the pointer

farmerpaul reviewed (approved):
Nice work on the loading state. I just had one last comment regarding CSS length units, but preapproving! 🚀

sultanofcardio committed:
M2-5579: Convert pixels to rem
